### PR TITLE
DOC: Events overview

### LINKED ISSUE #7215: Add a summary of build phases and core events
In the documentation, there is a lot of good detail about each individual event that happens (e.g.: https://www.sphinx-doc.org/en/master/extdev/appapi.html#sphinx-core-events). However, it can be hard to wrap your head around the high-level picture of what happens over the course of a whole build.

In a project we are working on @chrisjsewell gave a really nice description of the birds-eye view of the build process. I am wondering if this could be added into the docs somewhere, because I think it'd be useful.

Here's his description - does it look correct, and if so, would it be useful to add as a PR? Maybe to https://www.sphinx-doc.org/en/master/extdev/index.html#build-phases ?

```
1. event.config-inited(app,config)
2. event.builder-inited(app)
3. event.env-get-outdated(app, env, added, changed, removed)
4. event.env-before-read-docs(app, env, docnames)

for docname in docnames:
    5.  event.env-purge-doc(app, env, docname)
    if doc changed and not removed:
        6. source-read(app, docname, source)
        7. run parser: text -> docutils.document
        8. apply transforms (by priority): docutils.document -> docutils.document
        9. event.doctree-read(app, doctree)

10. event.env-updated(app, env)
11. event.env-check-consistency(app, env)

for docname in docnames:
    12. apply post-transforms (by priority): docutils.document -> docutils.document
    13. event.doctree-resolved(app, doctree, docname)

14. call builder

15. event.build-finished(app, exception)
```

### PR DESCRIPTION
**Pull request by choldgraf**
This PR adds a short overview of the events emitted during the build process, to help extension developers understand when events happen in relation to one another, and the overall build process.

I tried to incorporate all of the feedback that @tk0miya gave, though likely this will need some more iteration. What do people think?

note: @tk0miya you mentioned an event "`get-env-updated`" but I couldn't find it documented in https://www.sphinx-doc.org/en/master/extdev/appapi.html

closes #7215

### CODE REVIEW

**Comment by choldgraf** [comment]
whoahh I see a ton of conflicting files here...perhaps I am making this PR against the wrong branch?

File: doc/extdev/index.rst [hunk 1]
```diff
@@ -189,6 +189,89 @@ as metadata of the extension.  Metadata keys currently recognized are:
   usually don't negatively influence the process, this defaults to ``True``.
 
 
+Extension How-tos
+-----------------
```
Comment (tk0miya) [inline doc/extdev/index.rst]: Is this change intended? It seems not related to "events overview".

**Comment by tk0miya** [comment]
It seems you added the changes into master branch. So I changed the target branch of this PR to master.

File: doc/extdev/appapi.rst [hunk 2]
```diff
@@ -156,6 +156,39 @@ connect handlers to the events.  Example:
...
+   11. event.env-get-updated(app, env)
+   11. event.env-check-consistency(app, env)
+
+   # For builders that output a single page, they are first joined into a single doctree before post-transforms/doctree-resolved
+   for docname in docnames:
+      12. apply post-transforms (by priority): docutils.document -> docutils.document
+      13. event.doctree-resolved(app, doctree, docname)
+          - (for any reference node that fails to resolve) event.missing-reference(env, node, contnode)
+
+   14. call builder
```
Comment (tk0miya) [inline doc/extdev/appapi.rst]: The builder controls the whole of the build. It starts to work just after "2. builder-inited" event to before "15. build-finished". I thought "generate output files" is better than this.

File: doc/extdev/index.rst (hunk 1 above)
Comment (choldgraf) [reply doc/extdev/index.rst]: nope, accident, I'll remove

**Comment by choldgraf** [comment]
oops! let me re-work this PR to be on 3.0 since that seems more appropriate

**Comment by choldgraf** [comment]
OK I just started fresh from `3.x` and switched the target of this PR to 3.x as well. Also addressed your comment about "Generate output files"

**Review by tk0miya — APPROVED** [review]
Excellent!

**Comment by choldgraf** [comment]
wohoo, thanks!
